### DISCUSSION TITLE
Is it possible to use Typescript/Javascript with Playwright to test C#/.Net applications?

### DISCUSSION THREAD

**Discussion by mruanGitHub** [reqnroll/Reqnroll#339] [Q&A] [answered]
As the title asking, or we must use C# to create step definition functions? So basically, could we use Typescript or Javascript to create step definition functions to run Playwright? which might be more easy or efficient compared to C#?
The system under test may be developed in C#/.Net.

**Comment by gasparnagy** [comment]
You don't need to. If you test your backend only through Web UI or API (with Playwright for example), you can use Cucumber.js to test it. See https://cucumber.io/docs/installation/javascript/ for details.

**Comment by mruanGitHub** [comment]
Thanks for reply. How about this:
"Cucumber is available for most mainstream programming languages. We recommend choosing an implementation for the same platform or programming language as the production code."
So the conclusion is that it is recommented to use C# for the step definitions, but still it is possible to use Typescript/Javascript (Cucumber.js)?

**Reply by gasparnagy** [reply, marked as answer]
No. That statement is emphasizing that in order to test your application with "domain automation" (creating instances of your production classes and call methods on them), you should use the same platform (so that would mean testing your C# backend with Reqnroll and C# step definitions). 

However if you do end-to-end or API testing of your project (as you have mentioned Playwright, I guess this is the case for you), you will never need to directly access your backend classes and methods, but only interact with your application through Playwright anyway, so in this case, you have the option to choose Cucumber.js with JavaScript/TypeScript step definitions. You cannot mix though: you cannot use Cucumber.js with C# step definitions or Reqnroll with JavaScript step definitions.
